File: twopointers/1782-count-pairs-of-nodes.go
package twopointers

import "sort"

// https://leetcode.com/problems/count-pairs-of-nodes/

// Approach: Two Pointers + Bitmask
// Time: O(q*(n+e) + nlogn), q=len(queries)
// Space: O(n+e), e=len(edges)
func countPairs(n int, edges [][]int, queries []int) []int {
	var degree = make([]int, n+1)
	// max(n) = 2^4, dec(2e5) = bin(0b100111000100000), # of bits = 15
	// so, we'll move the first node by 15 steps to left
	var shared = make(map[int]int)
	var u, v int
	for idx := range edges { // O(e)
		u, v = edges[idx][0], edges[idx][1]
		degree[u]++
		degree[v]++
		if u > v {
			u, v = v, u
		}
		shared[u<<15+v]++
	}
	var sortedDegree = make([]int, n)
	copy(sortedDegree, degree[1:]) // [1:] means excluding node 0
	sort.Ints(sortedDegree)        // O(nlogn)
	var answer = make([]int, len(queries))
	var left, right int
	for idx := range queries { // O(q)
		left, right = 0, n-1
		for left <= right { // O(n)
			if sortedDegree[left]+sortedDegree[right] > queries[idx] {
				answer[idx] += right - left
				right--
			} else {
				left++
			}
		}
		for mask := range shared { // O(e)
			u, v = mask>>15, mask&(1<<15-1)
			if degree[u]+degree[v] > queries[idx] && degree[u]+degree[v]-shared[u<<15+v] <= queries[idx] {
				answer[idx]--
			}
		}
	}
	return answer
}

// // Approach: Brute-Force
// // Time: O(q*n^2), q=len(queries)
// // Space: O(n+e), e=len(edges)
// func countPairs(n int, edges [][]int, queries []int) []int {
// 	var degree = make([]int, n+1)
// 	var shared = make([][]int, n+1)
// 	for idx := range shared { // O(n)
// 		shared[idx] = make([]int, n+1)
// 	}
// 	var u, v int
// 	for idx := range edges { // O(e)
// 		u, v = edges[idx][0], edges[idx][1]
// 		degree[u]++
// 		degree[v]++
// 		if u > v {
// 			u, v = v, u
// 		}
// 		shared[u][v]++
// 	}
// 	var answer = make([]int, len(queries))
// 	for idx := range queries { // O(q)
// 		for u := 1; u <= n; u++ { // O(n)
// 			for v := u + 1; v <= n; v++ { // O(n)
// 				if degree[u]+degree[v]-shared[u][v] > queries[idx] {
// 					answer[idx]++
// 				}
// 			}
// 		}
// 	}
// 	return answer
// }

File: twopointers/1782-count-pairs-of-nodes.py
from collections import defaultdict
from typing import List
import unittest

# https://leetcode.com/problems/count-pairs-of-nodes/

# python3 -m unittest twopointers/1782-count-pairs-of-nodes.py


class Solution(unittest.TestCase):
    # # Approach: Brute-Force
    # # Time: O(q*n^2), q=len(queries)
    # # Space: O(n+e), e=len(edges)
    # def countPairs(self, n: int, edges: List[List[int]], queries: List[int]) -> List[int]:
    #     degree = defaultdict(int)
    #     shared = defaultdict(int)
    #     for u, v in edges:
    #         degree[u] += 1
    #         degree[v] += 1
    #         shared[(min(u,v),max(u,v))] += 1
    #     answer = [0] * len(queries)
    #     for idx, query in enumerate(queries):
    #         for u in range(1, n+1):
    #             for v in range(u+1, n+1):
    #                 if degree[u] + degree[v] - shared[(u,v)] > query:
    #                     answer[idx] += 1
    #     return answer

    # Approach: Two Pointers
    # Time: O(q*(n+e) + nlogn), q=len(queries)
    # Space: O(n+e), e=len(edges)
    def countPairs(self, n: int, edges: List[List[int]], queries: List[int]) -> List[int]:
        degree = [0] * (n + 1)
        shared = defaultdict(int)
        for u, v in edges:  # O(e)
            degree[u] += 1
            degree[v] += 1
            shared[(min(u, v), max(u, v))] += 1
        answer = [0] * len(queries)
        sorted_degree = sorted(degree[1:])  # O(nlogn), [1:] means excluding node 0
        for idx, query in enumerate(queries):  # O(q)
            left, right = 0, n - 1
            while left <= right:  # O(n)
                if sorted_degree[left] + sorted_degree[right] > query:
                    answer[idx] += right - left
                    right -= 1
                else:
                    left += 1
            for (u, v), cnt in shared.items():  # O(e)
                # subtract those edges that were added, but
                # considering shared ones between them breaks condition
                if degree[u] + degree[v] > query >= degree[u] + degree[v] - cnt:
                    answer[idx] -= 1
        return answer

    def test(self):
        for n, edges, queries, expected in [
            (4, [[1, 2], [2, 4], [1, 3], [2, 3], [2, 1]], [2, 3], [6, 5]),
            (5, [[1, 5], [1, 5], [3, 4], [2, 5], [1, 3], [5, 1], [2, 3], [2, 5]], [1, 2, 3, 4, 5], [10, 10, 9, 8, 6]),
        ]:
            output = self.countPairs(n, edges, queries)
            self.assertListEqual(expected, output, f"expected: {expected}, output: {output}")
